samples/bluetooth/mesh/chat/src/model_handler.c

```diff
@@ -0,0 +1,424 @@
+/******************************************************************************/
+/* Set up a repeating delayed work to blink the DK's LEDs when attention is
+ * requested.
+ */
+static struct k_delayed_work attention_blink_work;
+
+static void attention_blink(struct k_work *work)
+{
+	static int idx;
+	const uint8_t pattern[] = {
+		BIT(0) | BIT(1),
+		BIT(1) | BIT(2),
+		BIT(2) | BIT(3),
+		BIT(3) | BIT(0),
+	};
+	dk_set_leds(pattern[idx++ % ARRAY_SIZE(pattern)]);
+	k_delayed_work_submit(&attention_blink_work, K_MSEC(30));
+}
+
+static void attention_on(struct bt_mesh_model *mod)
+{
+	k_delayed_work_submit(&attention_blink_work, K_NO_WAIT);
+}
+
+static void attention_off(struct bt_mesh_model *mod)
+{
+	k_delayed_work_cancel(&attention_blink_work);
+	dk_set_leds(DK_NO_LEDS_MSK);
+}
+
+static const struct bt_mesh_health_srv_cb health_srv_cb = {
+	.attn_on = attention_on,
+	.attn_off = attention_off,
+};
+
+static struct bt_mesh_health_srv health_srv = {
+	.cb = &health_srv_cb,
+};
+
+BT_MESH_HEALTH_PUB_DEFINE(health_pub, 0);
+
+/******************************************************************************/
+/***************************** Chat model setup *******************************/
+/******************************************************************************/
+struct presence_cache {
+	uint16_t addr;
+	enum bt_mesh_chat_cli_presence presence;
+};
+
+/* Cache of the received Presence statuses of other Chat Client models. */
+static struct presence_cache presence_cache[
+			    CONFIG_BT_MESH_CHAT_SAMPLE_PRESENCE_CACHE_SIZE];
+
+static const uint8_t *presence_string[] = {
+	[BT_MESH_CHAT_CLI_PRESENCE_AVAILABLE] = "available",
+	[BT_MESH_CHAT_CLI_PRESENCE_AWAY] = "away",
+	[BT_MESH_CHAT_CLI_PRESENCE_DO_NOT_DISTURB] = "dnd",
+	[BT_MESH_CHAT_CLI_PRESENCE_INACTIVE] = "inactive",
+};
+
+/**
+ * Returns true if the specified address is an address of the local element.
+ */
+static bool address_is_local(struct bt_mesh_model *mod, uint16_t addr);
+
+/**
+ * Returns true if the node is new or the presence status is different from
+ * the one stored in the cache.
+ */
+static bool presence_cache_entry_check_and_get(uint16_t addr,
+				       enum bt_mesh_chat_cli_presence presence,
+				       size_t *entry)
+{
+	size_t i;
+
+	/* Find address in cache. */
+	for (i = 0; i < ARRAY_SIZE(presence_cache); i++) {
+		if (presence_cache[i].addr == addr) {
+			if (presence_cache[i].presence == presence) {
+				return false;
+			}
+
+			/* Break since the node in the cache. */
+			break;
+		}
+	}
+
+	/* Not in cache. */
+	if (i == ARRAY_SIZE(presence_cache)) {
+		for (i = 0; i < ARRAY_SIZE(presence_cache); i++) {
+			if (!presence_cache[i].addr) {
+				break;
+			}
+		}
+
+		/* Cache is full. */
+		if (i == ARRAY_SIZE(presence_cache)) {
+			/* Move all entries by 1. */
+			for (i = 1; i < ARRAY_SIZE(presence_cache); i++) {
+				presence_cache[i - 1] = presence_cache[i];
+			}
+
+			i = ARRAY_SIZE(presence_cache) - 1;
+		}
+	}
```

Comment:
Might be simpler to just keep a static `head` index that loops around the array instead of shifting the entire content